recipe/migrations/numpy2.yaml

```diff
@@ -0,0 +1,57 @@
+__migrator:
+  build_number: 1
+  kind: version
+  commit_message: |
+    Rebuild for numpy 2.0
+    
+    As of version 2.0, the way we build against numpy has changed. In particular,
+    we no longer need to use the oldest available numpy version at build time
+    in order to support old numpy version at runtime - numpy will by default
+    use a compatible ABI for the oldest still-supported numpy versions.
+    
+    Additionally, we no longer need to use `{{ pin_compatible("numpy") }}` as a
+    run requirement - this has been handled for more than two years now by a
+    run-export on the numpy package itself. The migrator will therefore remove
+    any occurrences of this.
+    
+    However, by default, building against numpy 2.0 will assume that the package
+    is compatible with numpy 2.0, which is not necessarily the case. You should
+    check that the upstream package explicitly supports numpy 2.0, otherwise you
+    need to add a `- numpy <2` run requirement until that happens (check numpy
+    issue 26191 for an overview of the most important packages).
+    
+    Finally, numpy 2.0 has not yet been released, but there is a release candidate
+    that promises to be ABI-compatible with the final release. This means we can
+    build against numpy 2.0.0rc1 (which can only be installed using a special channel
+    to opt in, i.e. `conda install -c conda-forge/label/numpy_rc numpy=2.0`), and yet
+    produce packages that are compatible with older numpy versions and therefore can
+    be published to our main channels.
```

Comment:
I'm suggesting to provide some more extensive information in the commit message (and thus PR body). I hope this will make things clearer for maintainers, but would like people's feedback here. @conda-forge/core @conda-forge/numpy 

PS. I'm intentionally not using the link to https://github.com/numpy/numpy/issues/26191, as otherwise, ~thousands of PRs would create a GH-visible link to that issue, which is just gratuitous spam from the POV of numpy.

Reply:
My main comment here is that it'd be nice to have a clear statement of what maintainers should do with this PR. The extra info made me wonder. I think something like the text below right at the top would help?

> TL;DR The way we build against numpy has changed as of numpy 2.0. This bot PR has updated the recipe to account for the changes (see below for details). Once your code is compatible with numpy 2.0 and this PR passes, you can merge it. The numpy 2.0 package itself is on a special release channel (`conda-forge/label/numpy_rc`) and will not be available on the main `conda-forge` channel until we fully migrate. To use packages with numpy 2.0, add `conda-forge/label/numpy_rc` to the top of your channels in your `.condarc`.

Reply:
Both the detailed commit message proposed and the added tl;dr suggestion seem 👍🏼 to me.

Reply:
Yeah Matt raises a good point. Maybe we can add a checklist

```suggestion
  commit_message: |
    Rebuild for numpy 2.0
    
    As of version 2.0, the way we build against numpy has changed. In particular,
    we no longer need to use the oldest available numpy version at build time
    in order to support old numpy version at runtime - numpy will by default
    use a compatible ABI for the oldest still-supported numpy versions.
    
    Additionally, we no longer need to use `{{ pin_compatible("numpy") }}` as a
    run requirement - this has been handled for more than two years now by a
    run-export on the numpy package itself. The migrator will therefore remove
    any occurrences of this.
    
    However, by default, building against numpy 2.0 will assume that the package
    is compatible with numpy 2.0, which is not necessarily the case. You should
    check that the upstream package explicitly supports numpy 2.0, otherwise you
    need to add a `- numpy <2` run requirement until that happens (check numpy
    issue 26191 for an overview of the most important packages).
    
    Finally, numpy 2.0 has not yet been released, but there is a release candidate
    that promises to be ABI-compatible with the final release. This means we can
    build against numpy 2.0.0rc1 (which can only be installed using a special channel
    to opt in, i.e. `conda install -c conda-forge/label/numpy_rc numpy=2.0`), and yet
    produce packages that are compatible with older numpy versions and therefore can
    be published to our main channels.

    Todos:
    * [ ] Confirm `{{ pin_compatible("numpy") }}` is removed
    * [ ] Confirm upstream has NumPy 2 compatibility or if not add `numpy <2` upper bound in `run`
```

Reply:
I think I've incorporated all input here, though I didn't take it 1:1 (e.g. I shortened the TL;DR and added the channel instructions further down). Please let me know what you think!